src/rt/test/func/bag/bag.cc

```diff
@@ -0,0 +1,84 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#include "ds/bag.h"
+
+#include "test/harness.h"
+#include "test/log.h"
+#include "verona.h"
+
+using namespace snmalloc;
+using namespace verona::rt;
+
+void test()
+{
+  auto& alloc = ThreadAlloc::get();
+
+  Bag<uintptr_t, uintptr_t, Alloc> bag;
+
+  const int NUM_OBJECTS = 128;
+
+  // Allocate a bunch of objects so that the region vector uses many
+  // allocation blocks to track each object in the region.
+  for (uintptr_t i = 0; i < NUM_OBJECTS; i++)
+  {
+    bag.insert({nullptr, i}, alloc);
+  }
+
+  // Iterate through the bag and make sure that all the
+  // entries make sense.
+  Bag<uintptr_t, uintptr_t, Alloc>::iterator iter(&bag);
+
+  Bag<uintptr_t, uintptr_t, Alloc>::Elem* rm1 = nullptr;
+  Bag<uintptr_t, uintptr_t, Alloc>::Elem* rm2 = nullptr;
+  Bag<uintptr_t, uintptr_t, Alloc>::Elem* rm3 = nullptr;
+
+  uintptr_t count = 0;
+  for (auto entry : iter)
+  {
+    assert(entry->metadata == (NUM_OBJECTS - count - 1));
+
+    if (count == 5)
+      rm1 = entry;
+
+    if (count == 40)
+      rm2 = entry;
+
+    if (count == 100)
+      rm3 = entry;
+    count++;
+  }
+
+  assert(count == NUM_OBJECTS);
+
+  bag.remove(rm3);
+  bag.remove(rm2);
+  bag.remove(rm1);
+
+  // Check that the iterator skips over holes left in the region vector
+  count = 0;
+  for (auto item : iter)
+  {
+    UNUSED(item);
+    count++;
+  }
+  assert(count == NUM_OBJECTS - 3);
+
+  // Check that the freelist is used to fill existing holes before bump
+  // allocation.
+  auto idx1 = bag.insert({nullptr, 123}, alloc);
+  assert((uintptr_t)rm1 == (uintptr_t)idx1);
+
+  auto idx2 = bag.insert({nullptr, 456}, alloc);
+  assert((uintptr_t)rm2 == (uintptr_t)idx2);
+
+  auto idx3 = bag.insert({nullptr, 789}, alloc);
+  assert((uintptr_t)rm3 == (uintptr_t)idx3);
+
```

Comment:
Can you allocate one more that won't be using free list at this point to get the coverage for changing mode.

Reply:
Added as part of 57cbf756da87743cd0237dfb6989bff8a9462279